README.md
# Instagram Basic Display API
Instagram Basic Display API is an HTTP-based API that Facebook apps can use to get an Instagram user's profile, images, videos, and albums. This module provides an easy way to access that data.

## Requirements
* ProcessWire >= 2.7
* A Facebook Developer account
* Access to an Instagram user account

## Creating a Facebook App

These instructions will assist you in creating a Facebook App for the Instagram Basic Display API.

The app you will create uses the [*User Token Generator*](https://developers.facebook.com/docs/instagram-basic-display-api/overview#user-token-generator) for authentication - it does not need to be submitted for App Review. However, this does mean that you need to be able to login to the Instagram account for approval. If setting up for a client, suggest arranging a time when they can change the password temporarily to allow you to access the account and authenticate the app.

### Create a New Facebook App
1. Login to your account at [https://developers.facebook.com/](https://developers.facebook.com/).
2. *My Apps > Create App*.
3. *Select an app type* - Make sure to choose **Consumer** ! Otherwise you won't be able to choose the correct *product* (see point 7).
4. Give your app a *Display Name*. This cannot contain certain reserved words such as `Instagram`, `IG`, `Insta` or `Facebook`. I recommend using something like `Image Feed - ProcessWire Website`.
5. Add a *App Contact Email* if not already populated.
6. Click **Create App** button and complete the Security Check if required.
7. You will be prompted to *Add a Product*. Find *Instagram Basic Display* and click **Set Up**.

### Configure the Facebook App
You should now see *Instagram* listed in *Products* in the left sidebar. Before continuing with the Basic Display setup, you need to configure the app itself.

#### Settings > Basic
1. Scroll to the bottom of this page and click **Add Platform**.
2. Select **Website**.
3. Enter the URL of the site the module will be used on. If you intend to use this module on multiple sites, use your company's own website.
4. **Save Changes**

You can add any other information to this screen that you wish, such as an *App Icon*.

#### Settings > Advanced (Optional)
1. *Security > Server IP Whitelist* - Whitelist any server IPs here.
2. *Business Manager* - If you have a Business Manager account, select it here to assign ownership of the app.

### Basic Display Configuration
Now your Facebook app is ready, you can configure it to access the Instagram Basic Display API.

#### Create an Instagram Basic Display App
*Instagram > Basic Display*, click **Create New App** then **Create App**.

The app requires some settings:
* *Client OAuth Settings* - Add `https://{your-website.com}/{your-admin-url}/InstagramBasicDisplayApi` as a *Valid OAuth Redirect URI*.
* *Deauthorize Callback URL* - Set this to `https://{your-website.com}/{your-admin-url}/InstagramBasicDisplayApi?action=deauthorize`.
* *Data Deletion Request URL* - Set this to `https://{your-website.com}/{your-admin-url}/InstagramBasicDisplayApi?action=delete`.

The Redirect URI should never be called by the API. However, if a user does deauthorize your app, or request deletion, the module is setup to handle these requests and notify you, so it makes sense to set actual URLs here.

#### Add the Instagram user account
1. *Roles > Instagram Testers*.
2. Click **Add Instagram Testers**.
3. Enter the username of the account to be added and click it when it appears.
4. Click **Submit**.
5. Login to the Instagram Account through [instagram.com](https://instagram.com/).
6. *Settings > Apps and Websites*. On the website, click the username to go to the profile, then click the "cog" icon.
7. Click on **Tester Invites** then click **Accept** for the app you have created.
8. Back in the App Dashboard, go to *Instagram > Basic Display > User Token Generator*.
9. You should now see the user account here. Click on **Generate Token**.
10. Log in to the Instagram Account again if necessary or click the continue button.
11. **Authorize** the app.
12. Check *I Understand* then copy the token that has been generated. Save it somewhere as you will need it to configure the module.

These instructions were current as of March 2020. If you notice any changes to how the Facebook Developer platform operates, please let me know on the [Support Forum](https://processwire.com/talk/topic/23028-instagrambasicdisplayapi/).

## Module Configuration
After you have installed this module:
1. *Modules > Configure > InstagramBasicDisplayApi*.
2. Add the username and generated token to *Add an Instagram User Account*.
3. Click **Submit**.
4. If successful, you should now see the account information in *Authorized Accounts*.

You can add as many accounts as you wish, as long as they have a token generated in your app.

The access token will be renewed automatically within the week prior to expiry.

## Methods

#### **getProfile(**_string_ **$username)**
Get a user's profile information.

This method returns an associative array with the following fields/keys:
- `username`
- `id`
- `account_type`
- `media_count`

```php
$instagram = $modules->get('InstagramBasicDisplayApi');

// Get the profile data of the default (first) user
$profile = $instagram->getProfile();

// Get the profile data of a specified user
$profile = $instagram->getProfile('username');

// Display the profile information
if(count($profile)) {
    $info = '';
    foreach($profile as $key => $value) {
        $info .= "<li>$key: $value</li>";
    }
    echo "<ul>$info</ul>";
}
```

#### **getImages(**_string_ **$username**, _int_ **$limit)**
Get a list of Images for a user.

This method returns a `WireArray` of `WireData` objects each with the following properties:
* `id` - The Media's ID.
* `type` - The Media's type. Can be *IMAGE*, *VIDEO* or *CAROUSEL_ALBUM*.
* `alt` - The Media's caption text. Not returnable for Media in albums.
* `description` - Alias of `alt`.
* `src` - The Media's URL.
* `url` - Alias of `src`.
* `tags` - An array of hashtags.
* `created` - The Media's publish date as a unix timestamp.
* `createdStr` - The Media's publish date in ISO 8601 format.
* `link` - The Media's permanent URL.
* `href` - Alias of `link`.

```php
$instagram = $modules->get('InstagramBasicDisplayApi');

// Get images from the default user
$images = $instagram->getImages(); // Returns all images found in the first request

// Get 10 images from the default user
$images = $instagram->getImages(10);

// Get images from a specified user
$images = $instagram->getImages('username'); // Returns all images found in the first request

// Get 8 images from a specified user
$images = $instagram->getImages('username', 8);

// Render the images
echo '<ul>' .
	$images->each('<li>' .
		'<a href={href}>' .
			'<img src={src} alt="{alt}">' .
		'</a>' .
	'</li>') .
'</ul>';
```

The main image of a carousel album and "poster" of a video is also returned.

#### **getCarouselAlbum(**_string_ **$username)**
Get the most recent Carousel Album for a user.

This method returns a `WireData` object with the same properties as `getImage()`. It also has an additional `children` property which contains a `WireArray` of the album's images.

```php
$instagram = $modules->get('InstagramBasicDisplayApi');

// Get the most recent album from the default user
$album = $instagram->getCarouselAlbum();

// Get the most recent album from a specified user
$album = $instagram->getCarouselAlbum('username');

// Render the album
if(isset($album)) {
	if(!$album->children) return '';
	echo '<ul>' .
		$album->children->each('<li>' .
			'<a href={href}>' .
				'<img src={src} alt="{alt}">' .
			'</a>' .
		'</li>') .
	'</ul>';
}
```

#### **getCarouselAlbums(**_string_ **$username**, _int_ **$limit)**
Get a list of Carousel Albums for a user.

This method returns a `WireArray` of `WireData` objects each with the same properties as `getImage()`. Each item also has an additional `children` property which contains a `WireArray` of the album's images.

This method should be used with care, as many API calls may need to be made to find the carousel albums requested. It is recommended to only use this if the Instagram user posts carousel albums frequently.

```php
$instagram = $modules->get('InstagramBasicDisplayApi');

// Get albums from the default user
$albums = $instagram->getCarouselAlbums(); // 4 returned if found

// Get 2 albums from the default user
$albums = $instagram->getCarouselAlbums(2);

// Get albums from a specified user
$albums = $instagram->getCarouselAlbums('username'); // 4 returned if found

// Get 3 albums from a specified user
$albums = $instagram->getCarouselAlbums('username', 3);

// Render the albums
if($albums->count()) {
	echo '<ul>' .
		$instagram->getCarouselAlbums()->each(function($album) {
			if(!$album->children) return '';
			return '<li>' .
				'<ul>' .
					$album->children->each('<li>' .
						'<a href={href}>' .
							'<img src={src} alt="{alt}">' .
						'</a>' .
					'</li>') .
				'</ul>' .
			'</li>';
		}) .
	'</ul>';
}
```

#### **getVideo(**_string_ **$username)**
Get the most recent Video for a user.

This method returns a `WireData` object with the same properties as `getImage()`. There is also an additional property for this media type:
- `poster` - The Media's thumbnail image URL.

```php
$instagram = $modules->get('InstagramBasicDisplayApi');

// Get the most recent video from the default user
$video = $instagram->getVideo();

// Get the most recent video from a specified user
$video = $instagram->getVideo('username');

// Render the video
if(isset($video)) {

	echo '<video ' .
		"src=$video->src " .
		"poster=$video->poster " .
		'type=video/mp4 ' .
		'controls ' .
		'playsinline' .
	'></video>';

	if($video->description) {
		echo "<p>$video->description</p>";
	}
}
```

#### **getVideos(**_string_ **$username**, _int_ **$limit)**
Get a list of Videos for a user.

This method returns a `WireArray` of `WireData` objects each with the same properties as `getImage()`. There is also an additional property for this media type:
- `poster` - The Media's thumbnail image URL.

This method should be used with care, as many API calls may need to be made to find the videos requested. It is recommended to only use this if the Instagram user posts videos frequently.

```php
$instagram = $modules->get('InstagramBasicDisplayApi');

// Get videos from the default user
$videos = $instagram->getVideos(); // 4 returned if found

// Get 2 videos from the default user
$videos = $instagram->getVideos(2);

// Get videos from a specified user
$videos = $instagram->getVideos('username'); // 4 returned if found

// Get 3 videos from a specified user
$videos = $instagram->getVideos('username', 3);

// Render the videos
if($videos->count()) {
	echo '<ul>' .
		$videos->each('<li>' .
			'<video ' .
				'src="{src}" ' .
				'poster="{poster}" ' .
				'type="video/mp4" ' .
				'controls ' .
				'playsinline' .
			'></video>' .
		'</li>') .
	'</ul>';
}
```

#### **getMedia(**_string_ **$username**, _int_ **$limit)**
Get a list of Media for a user.

The following example demonstrates how this can be used to create a multi-media gallery using the [UIkit](https://getuikit.com) [Grid](https://getuikit.com/docs/grid), [Cover](https://getuikit.com/docs/cover) and [Lightbox](https://getuikit.com/docs/lightbox) components:

```php
// Function for rendering items
function renderInstagramItem($src, $alt, $href = null) {
	if(is_null($href)) $href = $src;
	return "<a href=$href data-caption='$alt' class='uk-display-block uk-cover-container'" . ($src !== $href ? " data-poster=$src" : '') . ">" .
		'<canvas width=640 height=640></canvas>' .
		"<img src=$src alt='$alt' data-uk-img data-uk-cover>" .
	'</a>';
}

// Get the module
$instagram = $modules->get('InstagramBasicDisplayApi');

// Get the 16 most recent items and render them based on type
$items = [];
foreach($instagram->getMedia(16) as $item) {
	switch($item->type) {
		case 'VIDEO':
			$items[] = renderInstagramItem($item->poster, $item->alt, $item->src);
			break;
		case 'CAROUSEL_ALBUM':
			// If 4 or greater items, display a grid of the first 4 images
			// Otherwise display the main image (no break, moves to default)
			if($item->children->count() >= 4) {
				$out = '';
				$i = 0;
				foreach($item->children as $child) {
					$out .= '<div' . ($i++ < 4 ? '' : ' class="uk-hidden"') . '>' . // Hides items after the 4th one
						renderInstagramItem($child->src, $item->alt) .
					'</div>';
				}
				$items[] = "<div class='uk-grid-collapse uk-child-width-1-2' data-uk-grid>$out</div>";
				break;
			}
		default: // IMAGE
			$items[] = renderInstagramItem($item->src, $item->alt);
			break;
	}
}

// Render the items as a grid
echo '<div class="uk-grid-small uk-child-width-1-2 uk-child-width-1-3@s uk-child-width-1-4@l" data-uk-grid data-uk-lightbox>';
foreach($items as $item) {
	echo "<div>$item</div>";
}
echo '</div>';
```

## Pagination
Lazy loading of items can be achieved using `getMedia()`. The following example shows how to return batches of items using AJAX requests:

#### PHP
```php
$instagram = $modules->get('InstagramBasicDisplayApi'); // The pagination cursor is reset if the request is not AJAX, e.g. when the page is loaded.

if($config->ajax) {
	header('Content-Type: application/json');
	echo $instagram->getMedia(); // ['json' => true] is inferred by $config->ajax
	die();
}

echo '<div id=instagram class="uk-grid-small uk-child-width-1-2 uk-child-width-1-3@s uk-child-width-1-4@l" data-uk-grid data-uk-lightbox data-uk-scrollspy="target: > div; cls: uk-animation-slide-bottom-small; delay: 128;"></div>";

```

#### Javascript
```javascript
const instagram = {

	$el: {}, // Where the items go
	$loading: {}, // The loading spinner
	total: 0, // The total number of items

	init: () => {

		instagram.$el = UIkit.util.$('#instagram');
		if (!instagram.$el) return;

		// Add the spinner
		UIkit.util.after(instagram.$el, '<div id=instagram-loading class="uk-text-center uk-margin-top uk-hidden"><span data-uk-spinner></span></div>');
		instagram.$loading = UIkit.util.$('#instagram-loading');

		// Get the first batch of items
		instagram.get();
	},

	get: function() {

		// Show spinner
		UIkit.util.removeClass(instagram.$loading, 'uk-hidden');

		// Request
		UIkit.util.ajax(window.location.href, {
			method: 'GET',
			headers: {'X-Requested-With': 'XMLHttpRequest'},
			responseType: 'json'
		}).then(
			xhr => {

				// Hide spinner
				UIkit.util.addClass(instagram.$loading, 'uk-hidden');

				const data = xhr.response;
				if (!UIkit.util.isArray(data) || !data.length) return; // If no items do not render

				const items = [];
				data.forEach(item => {

					switch (item.type) {
						case 'VIDEO':
							items.push(instagram.renderItem(item.poster, item.alt, item.src));
							break;
						case 'CAROUSEL_ALBUM':
							// If 4 or greater items, display a grid of the first 4 images with the rest hidden
							// Otherwise display the main image (no break, moves to default)
							if (item.children.length >= 4) {
								let out = '';
								for (let i = 0; i < item.children.length; i++) {
									out += `<div${i < 4 ? '' : ' class=uk-hidden'}>
										${instagram.renderItem(item.children[i].src, item.alt)}
									</div>`;
								}
								items.push(`<div class="uk-grid-collapse uk-child-width-1-2" data-uk-grid>${out}</div>`);
								break;
							}
						default: // IMAGE
							items.push(instagram.renderItem(item.src, item.alt));
							break;
					}
				});

				const count = items.length;
				if (count) {

					// Wrap all items with a div
					let out = '';
					for (let i = 0; i < count; i++) {
						out += `<div id=instagram-item-${instagram.total + i}>
							${items[i]}
						</div>`;
					}

					// Append items to the container
					UIkit.util.append(instagram.$el, out);

					// Attach scrollspy listener on last item of second last row
					if (count > 5) {
						UIkit.util.on(`#instagram-item-${instagram.total + count - 6}`, 'inview', () => instagram.get());
					}

					// Update total
					instagram.total = instagram.total + count;
				}
			},
			e => {
				UIkit.util.addClass(instagram.$loading, 'uk-hidden');
				console.log(e); // ERROR
			}
		);
	},

	renderItem: (src, alt, href = '') => {
		if (href === '') href = src;
		return `<a href=${href} data-caption="${alt}" class="uk-display-block uk-cover-container"${src !== href ? ` data-poster=${src}` : ''}>
			<canvas width=640 height=640></canvas>
			<img src=${src} alt="${alt}" loading=lazy data-uk-cover>
		</a>`;
	}
};

UIkit.util.ready(() => instagram.init());
```
The javascript example uses [UIkit's Javascript Utilities](https://github.com/uikit/uikit-site/blob/feature/js-utils/docs/pages/javascript-utilities.md) but the same effect is achievable using plain javascript or any common library like jQuery.

## Backwards Compatibility with Instagram Feed
The primary reason for this module's development was to retain functionality on existing websites that use the [Instagram Feed](https://modules.processwire.com/modules/instagram-feed/) module, which uses a now deprecated API.

To assist with upgrading, this module replicates some methods provided by Instagram Feed:

#### **getRecentMedia(**_string_ **$username)**
Get the most recent media published by a user.

This is probably the most commonly used method from [InstagramFeed](https://modules.processwire.com/modules/instagram-feed/). You should not need to change your method call, however some of the values returned by the deprecated API are no longer present and are returned as `null` instead.

#### **getRecentMediaByTag(**_string_ **$tag**, _string_ **$username)**
Get a list of recently tagged media.

Instagram Basic Display API does not provide a way to search media by tag. This implementation will keep calling the API until it has enough matching items to return, or all media items have been retrieved.

Using this method is therefore **not recommended** as it is likely to slow response times and could possibly exhaust resource limits.

#### **setImageCount(**_int_ **$imageCount)**
Set the image count.

#### **getRecentComments(**_array_ **$media)**
Get recent comments. Returns a blank array as comments cannot be accessed by the Instagram Basic Display API.

#### **getUserIdByUsername(**_string_ **$username)**
Get the user's ID from their username.

### Upgrading to Instagram Basic Display Api
For a basic implementation, where all that Instagram Feed was requesting was the latest _n_ images, switching to use this module should only require installation and setup, then changing the module call:

```php
// Previously...
$instagram = $modules->get('InstagramFeed');
$images = $instagram->getRecentMedia();

// Now...
$instagram = $modules->get('InstagramBasicDisplayApi');
$images = $instagram->getRecentMedia();
```

If you had changed the default image count in the module config, you may also need to call `setImageCount()` to retain this number:
```php
// Images to return had been set to 8 in the InstagramFeed config
$instagram = $modules->get('InstagramBasicDisplayApi');
$images = $instagram->setImageCount(8)->getRecentMedia();
```

If you were getting images by tag, this also should work, but as noted above this can require numerous calls to the API, so should be used only if necessary and thoroughly tested to ensure there is no major impact on response times etc:
```php
$instagram = $modules->get('InstagramBasicDisplayApi');
$images = $instagram->getRecentMediaByTag('tagname');
```

As the Basic Display API differs from the deprecated one, there will be differences in what data is returned, but for the most part this should only be some values returning as `null` where they had a value previously such as `user_has_liked` or `users_in_photo`.

You should test thoroughly before deploying the upgrade.

## Notes
* By default, 24 media items are returned in each API call. You can change this in the module config. Up to 10K items can be returned at once, but it is not recommended!
* For methods related to InstagramFeed, e.g. `getRecentMedia()`, 4 items are returned by default, as this is the default in this module.
* Examples shown above may not work on older ProcessWire versions.


CHANGELOG.md
# Changelog

## 1.4.4 (October 19, 2022)

### Fixed
- `filterMedia()` now returning the video poster for the IMAGE type as it should.

### Changed
- Minor refactoring to reflect recent PW conventions, shouldn't impact any functionality.
- Javascript example in README updated to reflect current JS standards.


## 1.4.3 (September 2, 2021)

### Fixed
- Added unicode flag to preg_match_all to allow special characters in `tags`.


## 1.4.0 (June 27, 2020)

### Changed
- Applied a number of micro optimisations (e.g. single quotes where possible).


## 1.3.3 (April 15, 2020)

### Changed
- **InstagramBasicDisplayApi.module** Cache name now using `md5()` and not `base64_encode()`.


## 1.3.2 (April 9, 2020)

### Changed
- `account_type` re-enabled as the API error has been resolved by Facebook.


## 1.3.1 (April 7, 2020)

### Added
- **InstagramBasicDisplayApi.module** The `children` option for `getMedia()` can now be specified as an integer or string to use as a separate cache time for these API requests.

### Changed
- `account_type` disabled as the API throws an error incorrectly when this is requested.


## 1.3.0 (March 27, 2020)
Partial rewrite after the discovery of the `limit` API param. Everything should work as before, in fact it should work better!

### Added
- **InstagramBasicDisplayApiConfig.module** Added `limit` config option.
- **InstagramBasicDisplayApi.module** `getMedia()` added `children` option. If set to `false` carousel album children will not be requested/returned.
- **InstagramBasicDisplayApi.module** `getMedia()` added `json` option. If set to `true` a JSON string is returned. The default is equivalent to `$config->ajax` as generally when an AJAX request is made, JSON is the required response.

### Changed
- **InstagramBasicDisplayApi.module** `count` option for `getMedia()` changed name to `limit` and this now is implemented in the request.
- **InstagramBasicDisplayApi.module** Cache now only operates on `apiRequest()`.

### Fixed
- **InstagramBasicDisplayApi.module** `getMedia()` now removes items without a `media_url`.
- **InstagramBasicDisplayApi.module** `filterMedia()` now checks if a caption is present before trying to parse tags.

### Removed
- **InstagramBasicDisplayApi.module** `getCacheKey()` and `getCacheTime()` and moved their logic into `apiRequest()` as this is now the only place where cache is used.


## 1.2.1 (March 21, 2020)

### Added
- **README.md** `$instagram = $modules->get("InstagramBasicDisplayApi");` added to examples for clarity.
- **README.md** `getMedia()` example added.

### Removed
- **InstagramBasicDisplayApi.module** `getEndpoint()` and `urlGraph` constant - This is now handled in `apiRequest()` directly.
- **InstagramBasicDisplayApi.module** `refreshAccessToken()` - This is now handled in `apiRequest()` directly.


## 1.2.0 (March 12, 2020)
Module recoded to use the User Token Generator instead of OAuth.


## 1.0.1 (February 25, 2020)

### Fixed
- **InstagramBasicDisplayApi.module** `getMedia()` - `CAROUSEL_ALBUM` children now returned on all items of this type regardless of the type option.


InstagramBasicDisplayApiConfig.js
/**
 * Instagram Basic Display API Config JS
 *
 * @copyright 2022 NB Communication Ltd
 * @license Mozilla Public License v2.0 http://mozilla.org/MPL/2.0/
 *
 */

var InstagramBasicDisplayApiConfig = {

	init: function() {
		this.iconFix();
		this.trash();
	},

	iconFix: function() { // For PW 2.7.2

		var $spans = $('input.remove-account + span');
		if(!$spans.length) return;

		$spans.each(function() {
			var $span = $(this);
			if(!$span.children('i').length) {
				$span.html('<i class="fa fa-trash"></i>');
			}
		});
	},

	trash: function() {

		var $inputs = $('input.remove-account');
		if(!$inputs.length) return;

		$inputs.on('click', function() {
			var $input = $(this);
			var $row = $input.closest('tr');
			if($input.is(':checked')) {
				if(confirm('Are you sure you want to remove ' + $input.val() + "?\nThe account will be removed once you click Submit.")) {
					$row.addClass('removing-account');
				} else {
					$input.attr('checked', false);
				}
			} else {
				$row.removeClass('removing-account');
			}
		});
	}
};

$(document).ready(function() {
	InstagramBasicDisplayApiConfig.init();
});


InstagramBasicDisplayApiConfig.php
<?php

/**
 * Instagram Basic Display API Configuration
 *
 */

class InstagramBasicDisplayApiConfig extends ModuleConfig {

	/**
	 * Returns default values for module variables
	 *
	 * @return array
	 *
	 */
	public function getDefaults() {
		return [
			'cacheTime' => 3600,
			'limit' => 24,
		];
	}

	/**
	 * Returns inputs for module configuration
	 *
	 * @return InputfieldWrapper
	 *
	 */
	public function getInputfields() {

		$config = $this->wire()->config;
		$input = $this->wire()->input;
		$modules = $this->wire()->modules;
		$sanitizer = $this->wire()->sanitizer;
		$inputfields = parent::getInputfields();

		// Get the module
		$instagram = $modules->get(str_replace('Config', '', $this->className));

		if($input->post('_addUsername') && $input->post('_addToken')) {

			// Add user
			$username = $sanitizer->pageName(trim($input->post('_addUsername')));
			$token = $sanitizer->text(trim($input->post('_addToken')));
			if($username && $token && $instagram->addUserAccount($username, $token)) {
				$instagram->message(sprintf($this->_('%s has been added.'), $username));
			} else {
				$instagram->error(sprintf($this->_('Could not add user account %s'), $username));
			}

			$this->_addUsername = '';
			$this->_addToken = '';

		} else if($input->post('_removeAccount')) {

			// Remove user(s)
			foreach($input->post('_removeAccount') as $username) {
				$username = $sanitizer->pageName($username);
				if($username && $instagram->removeUserAccount($username)) {
					$instagram->message(sprintf($this->_('%s has been removed.'), $username));
				} else {
					$instagram->error(sprintf($this->_('Could not remove user account %s'), $username));
				}
			}

			$this->_removeAccount = [];

		} else if($input->post('clearCache')) {
			// Clear the Cache
			$this->wire()->cache->deleteFor($instagram);
			$this->message($this->_('Cache cleared'));
			$this->wire()->session->redirect($input->url(true));
		}

		// Add assets
		$urlAsset = "{$config->urls->root}site/modules/$instagram->className/{$this->className}";
		$config->styles->add("$urlAsset.css");
		$config->scripts->add("$urlAsset.js");

		// Get User Accounts
		$accounts = $instagram->getUserAccounts(true);

		// Add an Instagram User Account
		$fieldset = $modules->get('InputfieldFieldset');
		$fieldset->label = $this->_('Add an Instagram User Account');
		$fieldset->notes = implode("\n- ", [
			$this->_('Before adding a user you must first:'),
			sprintf($this->_('Add the user as an Instagram Tester in %s'), '*App Dashboard > Roles > Instagram Testers*'),
			sprintf($this->_('Get the user to approve the request in %s'), '*Settings > Apps and Websites*'),
			sprintf($this->_('Generate a User Token in %s'), '*App Dashboard > Products > Instagram > Basic Display > User Token Generator*'),
		]);
		$fieldset->icon = 'plus-circle';
		$fieldset->collapsed = count($accounts) ? 1 : 0;

		$fieldset->add([
			'type' => 'text',
			'name' => '_addUsername',
			'label' => $this->_('Username'),
			'value' => '',
			'icon' => 'instagram',
			'columnWidth' => 25,
		]);

		$fieldset->add([
			'type' => 'text',
			'name' => '_addToken',
			'label' => $this->_('Token'),
			'value' => '',
			'icon' => 'key',
			'columnWidth' => 75,
		]);

		$inputfields->add($fieldset);

		if(count($accounts)) {

			$datetime = $this->wire()->datetime;

			$table = $modules->get('MarkupAdminDataTable');
			$table->setEncodeEntities(false);
			$table->setSortable(false);
			$table->headerRow([
				$this->_('Username'),
				//$this->_('Account Type'),
				$this->_('User ID'),
				$this->_('Media Count'),
				$this->_('Token Renews'),
				'',
			]);

			foreach($accounts as $username => $account) {

				$id = "remove_$username";
				$remove = $modules->get('InputfieldCheckbox');
				$remove->attr('hidden', true);
				$remove->addClass('remove-account');
				foreach([
					'name' => '_removeAccount[]',
					'id' => $id,
					'value' => $username,
					'label' => wireIconMarkup('trash'),
					'entityEncodeLabel' => false,
					'labelAttrs' => ['for' => $id],
				] as $key => $value) {
					$remove->set($key, $value);
				};

				$renews = strtotime($account['token_renews']);
				$table->row([
					"<a href=https://www.instagram.com/$username target=_blank rel='noopener noreferrer'>$username</a>",
					//$account['account_type'],
					$account['user_id'],
					$account['media_count'],
					"<span title='$account[token_renews]'>" .
						(isset($datetime) ? $datetime->date('rel', $renews) : date('d/m/Y H:i', $renews)) .
					'</span>',
					$remove->render(),
				]);
			}

			$inputfields->add([
				'type' => 'markup',
				'id' => 'instagram-accounts',
				'label' => $this->_('Authorized Accounts'),
				'icon' => 'instagram',
				'value' => $table->render(),
			]);

			$inputfields->add([
				'type' => 'integer',
				'name' => 'limit',
				'label' => $this->_('Limit'),
				'description' => $this->_('The default number of items to return:'),
				'icon' => 'th',
				'collapsed' => 1,
			]);
		}

		// Cache Time
		$query = $this->wire()->database->prepare('SELECT name FROM caches WHERE name LIKE :name');
		$query->bindValue(':name', "{$instagram->className}__%");
		$query->execute();
		$numCaches = $query->rowCount();

		$inputfields->add([
			'type' => 'integer',
			'name' => 'cacheTime',
			'label' => $this->_('Cache'),
			'description' => sprintf(
				$this->_('The number of seconds an API response should be cached for. If set to %1$s or %2$s (1 week), a default cache time of %3$s will be used.'),
				'`0`',
				'>= `604800`',
				'`3600`'
			),
			'notes' => ($numCaches ? sprintf(
				$this->_n('There is currently %d cached request.', 'There are currently %d cached requests.',
				$numCaches
			), $numCaches) : ''),
			'icon' => 'files-o',
			'collapsed' => 1,
			'appendMarkup' => ($numCaches ?
				$modules->get('InputfieldSubmit')
					->attr('name+id', 'clearCache')
					->attr('value', $this->_('Clear Cache'))
					->render() :
				''
			),
		]);

		return $inputfields;
	}
}
